siocli/lib/SqlDump.cpp

```diff
@@ -0,0 +1,771 @@
+
+    // Create CodedInputStream only if row data is available to read
+    // otherwise codedInput constructor will be stucked on waiting on buffering data
+    google::protobuf::io::CodedInputStream codedInput(&input);
+
+    while (true) {
+        std::uint64_t rowLength = 0;
+        if (!codedInput.ReadVarint64(&rowLength)) std::runtime_error("Can't read row length");
+
+        if (rowLength == 0) break;
+        // nulls are dissalowed, read values
+        TableInfo tableInfo;
+        tableInfo.m_trid = readUint64(codedInput);
+        tableInfo.m_tableName = readString(codedInput);
+        tableInfo.m_currentColumnSetId = readUint64(codedInput);
+
+        tableInfos.push_back(std::move(tableInfo));
+    }
+
+    for (auto& tableInfo : tableInfos) {
+        tableInfo.m_columns = receiveColumnList(
+                connectionIo, input, databaseName, tableInfo.m_currentColumnSetId);
+
+        os << formCreateTableQuery(databaseName, tableInfo) << ';' << std::endl;
+    }
+
+    return tableInfos;
+}
+
+void dumpTableData(io::IoBase& connectionIo, std::ostream& os,
+        protobuf::CustomProtobufInputStream& input, const std::string& databaseName,
+        const TableInfo& table)
+{
+    std::stringstream ss;
+
+    // SELECT TRID,TYPE,NAME,CURRENT_COLUMN_SET_ID FROM <DATABASE>.SYS_TABLES WHERE NAME NOT LIKE 'SYS_%'"
+    ss << formSelectAllQuery(databaseName, table.m_tableName);
+
+    auto query = ss.str();
+    auto response = sendCommand(std::move(query), connectionIo, input);
+
+    std::vector<TableInfo> tableInfos;
+    tableInfos.reserve(32);
+
+    // + TRID
+    if (response.column_description_size() != static_cast<int>(table.m_columns.size() + 1))
+        std::runtime_error("Unexpected user table columns count");
+
+    bool nullAllowed = false;
+    for (int i = 0; i < response.column_description_size(); ++i) {
+        const auto& column = response.column_description(i);
+        nullAllowed |= column.is_null();
+    }
+
+    google::protobuf::io::CodedInputStream codedInput(&input);
+
+    std::uint64_t expectedTrid = 1;
+    while (true) {
+        std::uint64_t rowLength = 0;
+        if (!codedInput.ReadVarint64(&rowLength)) std::runtime_error("Can't read row length");
+
+        if (rowLength == 0) break;
+
+        stdext::bitmask nullBitmask;
+
+        if (nullAllowed) {
+            nullBitmask.resize(response.column_description_size(), false);
+
+            if (!codedInput.ReadRaw(nullBitmask.data(), nullBitmask.size())) {
+                std::ostringstream err;
+                err << "Can't read null bitmask from server: " << std::strerror(input.GetErrno());
+                throw std::system_error(input.GetErrno(), std::generic_category(), err.str());
+            }
+        }
+
+        const auto trid = readUint64(codedInput);
+        if (expectedTrid != trid) {
+            os << formAlterTableSetNextTrid(databaseName, table.m_tableName, trid) << ';'
+               << std::endl;
```

Comment:
I suggest '\n' since we only need to flush at the end of the whole dump. Please fix in other places too.